pages/api/configuration.ts

```diff
@@ -5,83 +5,127 @@ import { domainMiddleware, jwtVerifyMiddleware } from "../../lib/middlewares";
-  input
-    .filter(({ key }) => CONFIGURATION_KEYS.includes(key))
-    .map(({ key, value }) => ({ key, value }));
-
-const prepareResponseFromMetadata = (input: MetadataItem[]) => {
-  const output: MetadataInput[] = [];
-  for (const configurationKey of CONFIGURATION_KEYS) {
-    output.push(
-      input.find(({ key }) => key === configurationKey) ?? { key: configurationKey, value: "" }
-    );
-  }
-  return output.map(({ key, value }) => ({ key, value }));
-};
+import {
+  prepareMetadataFromRequest,
+  prepareResponseFromMetadata,
+} from "@/backend/configuration";
 
 const handler: NextApiHandler = async (request, response) => {
   let saleorDomain: string;
 
   try {
     saleorDomain = domainMiddleware(request) as string;
     await jwtVerifyMiddleware(request);
-  }
-  catch (e: unknown) {
+  } catch (e: unknown) {
     const error = e as MiddlewareError;
 
     console.error(error);
-    response
-      .status(error.statusCode)
-      .json({ success: false, message: error.message });
+    response.status(error.statusCode).json({
+      success: false,
+      error,
+    });
     return;
   }
-
-  const client = createClient(
-    `https://${saleorDomain}/graphql/`,
-    async () => Promise.resolve({ token: getAuthToken() }),
+  const client = createClient(`https://${saleorDomain}/graphql/`, async () =>
+    Promise.resolve({ token: getAuthToken() })
   );
 
   let privateMetadata;
+  const channelQuery = request.query.channel;
+  const channels =
+    typeof channelQuery === "string" ? [channelQuery] : channelQuery;
   switch (request.method!) {
     case "GET":
-      privateMetadata  = (
-        (await client.query<FetchAppMetadataQuery>(FetchAppMetadataDocument).toPromise()).data
-      )?.app?.privateMetadata!;
-
-      response.json({ success: true, data: prepareResponseFromMetadata(privateMetadata) });
+      if (!channels?.length) {
+        response.json({
+          success: false,
+          error: { message: "Query param 'channel' is required" },
+        });
+        return;
+      }
+      privateMetadata = (
+        await client
+          .query<FetchAppMetafieldsQuery, FetchAppMetafieldsQueryVariables>(
+            FetchAppMetafieldsDocument,
+            {
+              keys: channels,
+            }
+          )
+          .toPromise()
+      ).data?.app?.privateMetafields;
+      console.log("privateMetadata", privateMetadata);
+      if (privateMetadata) {
+        response.json({
+          success: true,
+          data: prepareResponseFromMetadata(privateMetadata, channels, true),
+        });
+      } else {
+        response.json({
+          success: false,
+          error: { message: "Unable to fetch app configuration." },
+        });
+      }
       break;
     case "POST":
+      if (!channels?.length) {
+        response.json({
+          success: false,
+          error: { message: "Query param 'channel' is required" },
+        });
+        return;
+      }
```

Comment:
This is repeated, we could make a func out of it or some sort of middleware, check app template or klaviyo app for reference.

Reply:
Yes, I would like to refactor api in another pull request.